### DISCUSSION TITLE
디렉터리 구조 변경 논의

### DISCUSSION THREAD

**Discussion by devkyoung2** [KakaoFunding/front-end#96] [General]
### 현재 디렉터리 구조
``` powershell
src
 ├─ assets  // 자산 파일
 ├─ components  // 컴포넌트
 │  ├─ ProductList  // 상품 리스트 컴포넌트
 │  └─ Header // 헤더 컴포넌트
 ├─ hooks  // 커스텀 훅
 ├─ mocks  // 목 데이터
 ├─ pages  // 라우팅 되는 페이지
 │  ├─ Product  // 상품 조회 페이지
 │  └─ CategoryResult  // 카테고리 결과 페이지
 ├─ services  // api 처리 로직
 ├─ stories  // 스토리
 ├─ styles  // 전역 스타일
 ├─ types  // 공용 타입
 └─ utils  // 공용 함수
```
개인적으로 느끼는 위 구조의 문제점입니다.

1. 한 페이지에서 사용되는 컴포넌트의 수가 많아지면 하나의 디렉터리에 많은 파일이 모여 복잡해진다.

    ![image](https://github.com/KakaoFunding/front-end/assets/79066587/bd1c0488-25ad-4023-8a91-0a8b79348244)

2. 컴포넌트 디렉터리에 특정 기능을 위한 컴포넌트(`ColumnProductItem`)와 모든곳에서 사용되는 공용 컴포넌트(`Button`) 이 모여있다.

    ![image](https://github.com/KakaoFunding/front-end/assets/79066587/85cac0d3-07f0-4355-95b6-05dcb4f97467)

### 논의하고 싶은 사항
1. feature 로 기능 컴포넌트들을 분리하는게 어떠신지 여쭤보고 싶습니다.
  1-1. 만약 반영된다면 feature를 src의 하위 카테고리로 분리하는것이 맞을까요 ?
2. 디렉터리 구조를 아래와 같이 바꾸는 것이 어떤지 여쭤보고 싶습니다.

### 개선된 디렉터리 구조
``` powershell
src
 ├─ assets  // 자산 파일
 ├─ components  // 컴포넌트
 │  ├─ feature // 기능 컴포넌트 🌟
 │  │   ├─ ColumnProductItem  // 상품 아이템 컴포넌트 🌟
 │  │   └─ ProductList  // 상품 리스트 컴포넌트 🌟
 │  └─ Header // 헤더 컴포넌트
 ├─ hooks  // 커스텀 훅
 ├─ mocks  // 목 데이터
 ├─ layouts  // 라우팅 되는 페이지 🌟
 │  └─ Product // 상품 조회 페이지에서 사용되는 컴포넌트들
 │        ├─ ProductDetail 🌟
 │        └─ ProductBuyInfo 🌟
 ├─ pages  // 라우팅 되는 페이지
 │  ├─ Product  // 상품 조회 페이지
 │  └─ CategoryResult  // 카테고리 결과 페이지
 ├─ services  // api 처리 로직
 ├─ stories  // 스토리
 ├─ styles  // 전역 스타일
 ├─ types  // 공용 타입
 └─ utils  // 공용 함수
```

**components**
- 공용으로 사용되는 컴포넌트만 이 곳에서 작성합니다.

**pages**
- 라우팅을 담당하는 페이지만 이 곳에서 작성합니다.
- 하나의 디렉터리에는 하나의 컴포넌트와 하나의 스타일 파일만 있습니다.

**layouts**
- 하나의 페이지에서만 사용되고 재사용 되지 않는 컴포넌트를 이 곳에 작성합니다.
- 트리구조에 알맞게 하위 디렉터리로 작성할 수 있습니다.

더 좋은 의견이 있다면 코멘트 부탁드립니다. 논의가 끝나면 바로 반영할 예정입니다!

**Comment by uraflower** [comment]
- pages/
- layout/
- components/feature/
- components/ui/ 
---
- 기존 App을 pages/로 이동
- Header, Footer를 `layout/App/`로 이동
---
branch 하나 파서 디렉토리 구조 변경 & 본인이 작업한 부분 디렉터리 구조 수정
